src/ElfUtils/ElfFile.cpp

```diff
@@ -65,8 +67,47 @@ class ElfFileImpl : public ElfFile {
   std::string build_id_;
   bool has_symtab_section_;
   bool has_debug_info_section_;
+  std::optional<GnuDebugLinkInfo> gnu_debuglink_info_;
 };
 
+template <typename ElfT>
+ErrorMessageOr<GnuDebugLinkInfo> ReadGnuDebuglinkSection(
+    const typename ElfT::Shdr& section_header, const llvm::object::ELFFile<ElfT>& elf_file) {
+  llvm::Expected<llvm::ArrayRef<uint8_t>> contents_or_error =
+      elf_file.getSectionContents(&section_header);
+  if (!contents_or_error) return ErrorMessage("Could not obtain contents.");
+
+  const llvm::ArrayRef<uint8_t>& contents = contents_or_error.get();
+
+  if (contents.size() < 4) return ErrorMessage{"Section is too short."};
+
+  if (contents.size() > 100 * 1024) {
+    return ErrorMessage{"Section is longer than 100K. Something is not right."};
+  }
+
+  std::string path{};
+  path.reserve(contents.size());
+
+  for (const auto& byte : contents) {
+    if (byte == '\0') break;
+    path.push_back(static_cast<char>(byte));
+  }
+
+  if (path.size() > contents.size() - 4) return ErrorMessage{"No CRC32 checksum found"};
+
+  constexpr llvm::support::endianness kEndianness = ElfT::TargetEndianness;
+  using ChecksumType = uint32_t;
+  const auto checksum_storage = contents.slice(contents.size() - sizeof(ChecksumType));
+  constexpr size_t kAlignment = 1;
+  const auto reference_crc =
+      llvm::support::endian::read<ChecksumType, kAlignment>(checksum_storage.data(), kEndianness);
```

Comment:
while this is technically correct, I do not believe big-endian architecture will see the light of day in observable future.

personal preference: static_assert for LE somewhere and forget about endianness

Reply:
Well there is MIPS, and technically ARM. :sweat_smile: 

But jokes aside, I see your point. Changed!